sui_core/src/gateway_types.rs

```diff
@@ -0,0 +1,984 @@
+                    return SuiMoveValue::ByteArray(Base64::from_bytes(&bytearray));
+                }
+                SuiMoveValue::Vector(value.into_iter().map(|value| value.into()).collect())
+            }
+            MoveValue::Struct(value) => {
+                // Best effort Sui core type conversion
+                if let MoveStruct::WithTypes { type_, fields } = &value {
+                    if let Some(value) = try_convert_type(type_, fields) {
+                        return value;
+                    }
+                };
+                SuiMoveValue::Struct(value.into())
+            }
+            MoveValue::Signer(value) => {
+                SuiMoveValue::Signer(SuiAddress::from(ObjectID::from(value)))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(untagged, rename = "MoveStruct")]
+pub enum SuiMoveStruct {
+    Runtime(Vec<SuiMoveValue>),
+    WithFields(BTreeMap<String, SuiMoveValue>),
+    WithTypes {
+        #[serde(rename = "type")]
+        type_: String,
+        fields: BTreeMap<String, SuiMoveValue>,
+    },
+}
+
+fn try_convert_type(type_: &StructTag, fields: &[(Identifier, MoveValue)]) -> Option<SuiMoveValue> {
+    let struct_name = format!(
+        "0x{}::{}::{}",
+        type_.address.short_str_lossless(),
+        type_.module,
+        type_.name
+    );
+    let fields = fields
+        .iter()
+        .map(|(id, value)| (id.to_string(), value.clone().into()))
+        .collect::<BTreeMap<_, SuiMoveValue>>();
+    match struct_name.as_str() {
+        "0x2::UTF8::String" | "0x1::ASCII::String" => {
+            if let SuiMoveValue::ByteArray(bytes) = fields["bytes"].clone() {
+                if let Ok(bytes) = bytes.to_vec() {
+                    if let Ok(s) = String::from_utf8(bytes) {
+                        return Some(SuiMoveValue::String(s));
+                    }
+                }
+            }
+        }
+        "0x2::Url::Url" => return Some(fields["url"].clone()),
+        "0x2::ID::ID" => {
+            if let SuiMoveValue::Address(id) = fields["bytes"] {
+                return Some(SuiMoveValue::ID(ID { bytes: id }));
+            }
+        }
+        "0x2::ID::UniqueID" => {
+            if let SuiMoveValue::ID(id) = fields["id"].clone() {
+                return Some(SuiMoveValue::UniqueID(UniqueID { id }));
+            }
+        }
+        "0x2::ID::VersionedID" => {
+            if let SuiMoveValue::UniqueID(id) = fields["id"].clone() {
+                if let SuiMoveValue::U64(version) = fields["version"].clone() {
+                    return Some(SuiMoveValue::VersionedID(VersionedID { id, version }));
+                }
+            }
+        }
+        "0x2::Balance::Balance" => {
+            if let SuiMoveValue::U64(value) = fields["value"].clone() {
+                return Some(SuiMoveValue::Balance(value));
+            }
+        }
+        "0x2::Coin::Coin" => {
```

Comment:
I noticed that the return type of `Coin` balance is number instead of string, can we make it always a string because JSON only supports up to 32 bits but balance can be 64 bits(the faucet uses U64_max)?